File: planka-apis/view-api/src/main/java/cn/agilean/kanban/api/view/request/ViewDataRequest.java
package dev.planka.api.view.request;

import dev.planka.domain.schema.definition.condition.Condition;
import lombok.Data;

import java.util.List;

/**
 * 视图数据查询请求
 */
@Data
public class ViewDataRequest {

    /**
     * 当前页码（从0开始）
     */
    private Integer page;

    /**
     * 每页大小
     */
    private Integer size;

    /**
     * 用户自定义排序（优先级高于视图配置）
     */
    private List<UserSortField> sorts;

    /**
     * 用户额外过滤条件（与视图条件 AND 合并）
     */
    private Condition additionalCondition;

    /**
     * 关键词搜索（快速搜索，匹配标题等）
     */
    private String keyword;

    /**
     * 指定分组值（用于分组展开时查询特定分组的数据）
     */
    private String groupValue;

    /**
     * 创建默认请求
     */
    public static ViewDataRequest defaultRequest() {
        ViewDataRequest request = new ViewDataRequest();
        request.setPage(0);
        request.setSize(20);
        return request;
    }
}

File: planka-ui/src/utils/view-data-request.ts
import type { SortField } from '@/types/view'
import type { Condition, ConditionNode } from '@/types/condition'
import { NodeType } from '@/types/condition'

/**
 * 与视图数据接口 {@link cn.planka.api.view.request.ViewDataRequest} 对齐的请求体
 */
export function andMergeConditions(
  a: Condition | undefined,
  b: Condition | undefined,
): Condition | undefined {
  const ra = a?.root
  const rb = b?.root
  if (!ra && !rb) {
    return undefined
  }
  if (!ra) {
    return b
  }
  if (!rb) {
    return a
  }
  return {
    root: {
      nodeType: NodeType.GROUP,
      operator: 'AND',
      children: [ra, rb] as ConditionNode[],
    },
  }
}

export function toViewDataQueryBody(input: {
  page?: number
  size?: number
  groupValue?: string
  userSorts?: SortField[]
  additionalCondition?: Condition
}): Record<string, unknown> {
  const body: Record<string, unknown> = {}
  if (input.page !== undefined) {
    body.page = input.page
  }
  if (input.size !== undefined) {
    body.size = input.size
  }
  if (input.groupValue !== undefined) {
    body.groupValue = input.groupValue
  }
  if (input.additionalCondition?.root) {
    body.additionalCondition = input.additionalCondition
  }
  if (input.userSorts?.length) {
    body.sorts = input.userSorts.map((s) => ({
      fieldId: s.field,
      direction: s.direction,
    }))
  }
  return body
}
